### PR TITLE
bpo-36144: Add PEP 584 operators to collections.ChainMap

### PR DESCRIPTION
**Pull request by curtisbucher**
Update collections.ChainMap to include | and |= operators.

https://bugs.python.org/issue36144

### CODE REVIEW

**Comment by brandtbucher** [comment]
Thanks for your time @curtisbucher, and welcome to CPython! 😉

File: Lib/test/test_collections.py
Comment (brandtbucher) [inline Lib/test/test_collections.py]: @curtisbucher, CI failed because of trailing space here:

```suggestion
        cm2 |= d
```

**Review by brandtbucher — APPROVED** [review]
Looks good to me!

File: Lib/collections/__init__.py
Comment (gvanrossum) [inline Lib/collections/__init__.py:982]: Please add an extra blank line here.

File: Lib/test/test_collections.py
Comment (gvanrossum) [inline Lib/test/test_collections.py:258]: Please add an extra blank line.

File: Misc/NEWS.d/next/Library/2020-03-07-11-26-08.bpo-36144.FG9jqy.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2020-03-07-11-26-08.bpo-36144.FG9jqy.rst]: I'd add "(``|`` and ``|=``)" after "operators", since not everybody knows what PEP 584 is. :-)

File: Lib/collections/__init__.py
Comment (gvanrossum) [inline Lib/collections/__init__.py]: I'd use `reversed()`, like in `__iter__()` above.

**Comment by gvanrossum** [comment]
Before I finish reviewing this, I have to bring up a tricky issue. There's a subtle behavior (I can't even find it in the reference manual :-( ) for binary operators.

Consider a class B that implements for example `__add__` and `__radd__` (the same applies to all binary operators). Now consider a subclass C that overrides `__add__` and `__radd__`. Let `b = B()` and `c = C()`, and evaluate `b + c`. Surprise: this calls `C.__radd__` rather than `B.__add__` (first)!

The reason for this exception is that if we didn't have this special behavior (i.e., if this would follow the normal behavior of trying `B.__add__` first), it would be difficult for the subclass C to override the behavior of `+` completely. A reasonable implementation of `B.__add__` would look like this:
```py
class B:
    def __add__(self, other):
        if not isinstance(other, B):
            return NotImplemented
        # return an instance of B representing self+other
```
This would totally succeed for `b + c`, and return an instance of B. (And the reason for *that* is also subtle -- since Python allows subclass constructors to have a signature that differs from the superclass constructor, there's no safe way for B to create an instance of `self.__class__`. We went over this extensively earlier in the discussion about `dict | dict`. :-)

Anyway, because we have this exception, I have constructed an example that behaves unintuitively (I think). Consider:
```py
class C(ChainMap):
    def __ror__(self, other):
        return super().__ror__(other)

c1 = ChainMap()
c2 = C()

print(c1 | c2)  # C(ChainMap({}))    # !?
print(c2 | c1)  # C({})
```
However, if we don't override the `__ror__` method, the first result printed changes to
```py
print(c1 | c2)  # ChainMap({})
```
The output without overridden `__ror__` is easily understood: both calls invoke `ChainMap.__or__`, and it returns an instance of the class of `self`. But when we override `__ror__` in C (i.e. the way shown in the example) we find that `C.__ror__` is called, which then calls `ChainMap.__ror__`. And that has an (IMO) unexpected behavior, resulting in a subclass of `ChainMap` whose `maps[0]` is itself a `ChainMap`.

I'm not sure, but I think the only way out may be to define `__ror__` differently, in a way that's more symmetric with `__or__`.

Thoughts?

**Comment by brandtbucher** [comment]
The lack of symmetry bothers me a tiny bit too (see `Mapping` vs `MutableMapping` for the type of `other`). To appease your fun subclass, we'd need to avoid calling `other.copy()` in `__ror__`... which is tricky.

However, we would avoid these issues altogether with something like:

```py
     def __ior__(self, other):
         self.maps[0].update(other)
         return self

     def __or__(self, other):
         if isinstance(other, _collections_abc.Mapping):
             return self.__class__(other.copy(), *self.maps)
         return NotImplemented

     def __ror__(self, other):
         if isinstance(other, _collections_abc.Mapping):
             return self.__class__(self.maps[0].copy(), *self.maps[1:], other)
         return NotImplemented
```

Much easier to explain and reason about... the only issue is that we lose (part) of the symmetry between `x |= y` and `x = x | y`, as you mentioned in the BPO issue:

> I'm not so keen on that, because its special behavior can't be mimicked by `|=`.

But I think I'm okay with that. The only alternative I see would be to insert a copy of `other` to the front of the chain when `|=` is called, which I don't like (since it changes the "primary" mapping).

You may still end up with nested ChainMaps, but at least it's clear why it's happening and it happens consistently. It's also nice that in all cases of binary `|`, the left operand is "copied" and the right operand is referenced (and you always end up with the same number of maps)!

**Comment by gvanrossum** [comment]
The problem with how you write that is that it calls `self.__class__(...)` -- and who knows what the signature of the subclass `__init__()` or `__new__()` is. But maybe we could emulate this with `copy()`? We know that the subclass must have a public `maps` attribute, so we could do this:
```py
    def __or__(self, other):
        if isinstance(other, _collections_abc.Mapping):
            new = self.copy()
            # This makes a copy of self.dict[0], which we undo like this:
            new.maps[:1] = [other.copy(), self.maps[0]]
            return new
        return NotImplemented

    def __ror__(self, other):
        if isinstance(other, _collections_abc.Mapping):
            new = self.copy()
            new.maps.append(other)
            return new
        return NotImplemented
```
TBH the `isinstance()` checks bother me -- there are no such checks elsewhere in `ChainMap` (which is in generally pleasantly straightforward, with no method bodies more than four lines).

I think we need to let this marinate a bit more before we commit to a given semantics. (Sorry @curtisbucher!)

**Comment by gvanrossum** [comment]
Oh wait. The existing `ChainMap` code already calls `self.__class__()` in a few places (everywhere it needs a new instance: `copy()`, `new_child()`, `parents`. So strike that objection. (I am still a bit uncomfortable with the `isinstance()` checks, and also with making `a |= b` different from `a = a | b` in terms of how many maps you end up with.)

**Comment by brandtbucher** [comment]
```py
    def __ior__(self, other):
        self.maps[0] |= other
        return self

    def __or__(self, other):
        return self.__class__(self.maps[0] | other, *self.maps[1:])

    def __ror__(self, other):
        m = other | self.maps[-1]
        if len(self.maps) == 1:
            return self.__class__(m)
        return self.__class__(self.maps[0].copy(), *self.maps[1:-1], m)
```

No `isinstance`, same number of maps each time, no more than four lines. :wink:

**Comment by curtisbucher** [comment]
@brandtbucher does this solution solve the initial issue with `__ror__` that @gvanrossum brought up? If so can you explain how to me?

**Comment by brandtbucher** [comment]
@curtisbucher The main idea here is that we pretty thinly wrap the `|` and `|=` operators for one of our underlying mappings (either the first or last, depending on order).

We get around the reflected `super()` hiccup by letting `other` take care of its own copy and update (in its own `__or__` method). So now, the behavior is consistent:

```py
>>> from collections import ChainMap
>>> class C(ChainMap):
...     def __ror__(self, other):
...         return super().__ror__(other)
... 
>>> c1 = ChainMap()
>>> c2 = C()
>>> c1 | c2
C(ChainMap({}))
>>> c2 | c1
C(ChainMap({}))
```

And if you don't override `__ror__` in the subclass, the results compose just as you might expect them to:

```py
>>> from collections import ChainMap
>>> class C(ChainMap):
...     pass
... 
>>> c1 = ChainMap
>>> c2 = C()
>>> c1 | c2
ChainMap(C({}))
>>> c2 | c1
C(ChainMap({}))
```

I think the above behavior for this sort-of-pathological case is reasonable, and not very surprising.

**Comment by gvanrossum** [comment]
Ah, this just keeps snaking out of control. :-(

I'm not particularly happy that with this solution, `ChainMap() | ChainMap()` produces a ChainMap whose `maps` references another ChainMap.

I also accept that if we want `ChainMap.__or__` to ever defer to a `__ror__` method on `other`, we'll have to use `isinstance()`.

I keep thinking that `c1 | x` should just be `c1.add_child(x)`, but then the question is what to do about `x | c1` and `c1 |= x`. And also repeated use of this will produce very long chains.

I wonder if the original solution can be tweaked to make my corner case work, rather than going with a completely different approach?

**Comment by brandtbucher** [comment]
> I wonder if the original solution can be tweaked to make my corner case work, rather than going with a completely different approach?

If we always want a dict as our first mapping, replacing `other.copy()` with `dict(other)` in `__ror__` should be sufficient. Your original subclass example:

```py
>>> from collections import ChainMap
>>> class Subclass(ChainMap):
...     pass
... 
>>> class SubclassRor(ChainMap):
...     def __ror__(self, other):
...         return super().__ror__(other)
... 
>>> ChainMap() | ChainMap()
ChainMap({})
>>> ChainMap() | Subclass()
ChainMap({})
>>> ChainMap() | SubclassRor()
SubclassRor({})
>>> Subclass() | ChainMap()
Subclass({})
>>> Subclass() | Subclass()
Subclass({})
>>> Subclass() | SubclassRor()
Subclass({})
```

I think forcing a `dict` here is okay, since we're collapsing all of the maps anyways.

File: Lib/collections/__init__.py
Comment (gvanrossum) [inline Lib/collections/__init__.py]: I think this ought to be `self.maps[0] |= other` -- most other ChainMap methods also defer to the corresponding method on `self.maps[0]`.

File: Lib/collections/__init__.py
Comment (gvanrossum) [inline Lib/collections/__init__.py:971]: This could be `m = self.maps[0] | other` -- also deferring to the same operation on `self.maps[0]`. Then you could inline it in the line below.

File: Lib/collections/__init__.py
Comment (gvanrossum) [inline Lib/collections/__init__.py]: We can use `Mapping` instead of `MutableMapping` and then `dict(other)` instead of `other.copy()`, per the end of the main comment thread.

**Comment by gvanrossum** [comment]
OK, I think that this is probably the best we can do if we don't want the number of maps to increase, which is dangerous.

Another way to look at this: The ideas based on making `|` somehow equivalent to a call to `add_child` are trying to make a special operation (`add_child`) look regular (`|`), whereas the priority should be to make regular operations work as expected. In particular, we want to allow decent semantics for ChainMap when passed in to code expecting a Mapping (or a duck type for `dict`).

The latest proposal fulfills this desire, so let's go for that. (And I won't change my mind again. :-)

File: Lib/test/test_collections.py
Comment (gvanrossum) [inline Lib/test/test_collections.py:257]: Probably some more tests for the edge case I gave.

File: Lib/collections/__init__.py
Comment (brandtbucher) [inline Lib/collections/__init__.py]: Here you go, @curtisbucher:

```suggestion
        if isinstance(other, _collections_abc.Mapping):
            m = dict(other)
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]

File: Lib/collections/__init__.py
Comment (brandtbucher) [reply Lib/collections/__init__.py]: Ah, we crossed posts.

**Comment by rhettinger** [comment]
Are ``_abc_collections.Mapping`` and ``_abc_collections.MutableMapping`` going to be updated for PEP 584?  Ideally, *ChainMap()* would just inherit those behaviors, much as it currently does with the *update()* method.

If not, this PR should at least call the existing *update()* method rather than touching ``self.maps[0]`` directly.  Not only would this simplify implementation and maintenance, it would avoid breaking the ``DeepChainMap(ChainMap)`` example in the docs which shows how to alter the target component mappings simply by overriding ``__setitem__`` and ``__delitem__``.

**Comment by brandtbucher** [comment]
> Are _abc_collections.Mapping and _abc_collections.MutableMapping going to be updated for PEP 584?

We decided not to do that for a few reasons. I don't have the exact posts handy right now, but I recall the points that shot the idea down were:

- `Mapping`/`MutableMapping` have no `copy` equivalent for creating new instances (necessary for binary operator).
- Even if they did, that adding the operators (or even just the in-place one) would create compatibility issues for virtual subclasses.

> If not, this PR should at least call the existing update() method rather than touching self.maps[0] directly.

So, just to clarify, you'd prefer we spell it more like this?

```py
    def __ior__(self, other):
        self.update(other)
        return self

    def __or__(self, other):
        if isinstance(other, _collections_abc.Mapping):
            m = self.copy()
            m.update(other)
            return m
        return NotImplemented

    def __ror__(self, other):
        if isinstance(other, _collections_abc.Mapping):
            m = dict(other)
            m.update(self)
            return self.__class__(m)
        return NotImplemented
```

I think that sounds reasonable.

**Comment by gvanrossum** [comment]
My only concern with that is that `x.update(other)` iterates over the key/value pairs in `other` and inserts them into `x` one at a time. This is slow for two reasons. FIrst, if `x` is a ChainMap, it makes lots of calls to `.__setitem__()`. Second, if `other` is a ChainMap, its `.__iter__()` constructs a fresh dict containing all key/value pairs, making the exact same set of `.update()` calls that the current code is making, thus doubling the memory requirement.

I don't know whether ChainMap was designed for performance -- maybe the abstractions are more important than speed? However, ISTM that ChainMap is frequently cited as a good solution for various problems that would otherwise require explicitly merging dicts, in contexts where ChainMap's API beyond MutableMapping doesn't seem important (i.e. seemingly because it's assumed to be faster).

I also note that the DeepChainMap example is already missing quite a few things, e.g. `.popitem()`, `.pop()`, `.clear()` and `.copy()`.

PS. @brandtbucher: Maybe it's worth explaining the two reasons for not adding these methods to the Mapping ABCs in the PEP, e.g. under Rejected Ideas?

**Comment by brandtbucher** [comment]
> Maybe it's worth explaining the two reasons for not adding these methods to the Mapping ABCs in the PEP, e.g. under Rejected Ideas?

I agree, since it's come up at least twice now since acceptance. I'll have something up soon.

File: Lib/test/test_collections.py
Comment (curtisbucher) [reply Lib/test/test_collections.py:257]: Did we already decide for sure how we wanted to address the edge case? Does Brandt's solution solve that in a way we like?

Comment (gvanrossum) [reply Lib/test/test_collections.py:257]: The edge case should not have ChainMap instances in its `maps` attribute. Brandt's solution of using `dict(other)` will ensure this.

File: Lib/collections/__init__.py
Comment (curtisbucher) [reply Lib/collections/__init__.py:971]: After implementing this I ran into the same problem we had earlier with your edge case, where `.__ror__` is called rather than `.__or__`.  This means that if `other` is a ChainMap we will be given nested ChainMaps which I know we didn't like. So maybe leave it as is?

```py
>>> ChainMap() | ChainMap()
ChainMap(ChainMap({}))
```

**Comment by rhettinger** [comment]
I'll have more time to think this through sometime in the next days (presuming there is no rush to apply 584 to ChainMap).  In particular, I want to think about use cases and user's mental models of what ChainMap does.  Also, I want to reread the PEP now that it has been approved.

FWIW, I don't recall seeing anyone ever use *update()* on a *ChainMap*.   That wouldn't be the normal way to use it.  If a user still has a variable reference to ``maps[0]``, presumably they would update that mapping directly rather than through the chainmap instance.

Aspirationally, it would be great if the new methods were implemented in terms of the existing methods and that we make the fewest possible assumptions about the API of the underlying mappings.

One other thought: the intended way to connect to another mapping is the *new_child()* method.

File: Lib/collections/__init__.py
Comment (gvanrossum) [reply Lib/collections/__init__.py:971]: You're right. We should either stick to what we have, or go with Brandt's [solution](https://github.com/python/cpython/pull/18832#issuecomment-596910350) from his reply to Raymond.

**Review by gvanrossum — COMMENTED** [review]
This LGTM, but I would also be okay if we implemented Brandt's version in his [reply](https://github.com/python/cpython/pull/18832#issuecomment-596910350) to Raymond.

Let's not change anything until we hear more from @rhettinger. (Raymond: it would be nice if this made the next alpha, which is [scheduled](https://www.python.org/dev/peps/pep-0596/) for March 23.)

**Review by brandtbucher — APPROVED** [review]

**Comment by gvanrossum** [comment]
@curtisbucher Did you know you can run most tests locally? E.g. to run `test_collections`, type this in you shell (Mac, Windows or Linux), assuming you've built `python.exe` locally:
```
./python.exe -m test test_collections
```
You can also build the docs locally. I'll add tips to my informal [guide](https://paper.dropbox.com/doc/Contributing-to-CPython--Av5YHhojR_3ur9dePQmmI9AtAg-JlgnduI6kw9MJIaGPpN9G) to contributing.

**Comment by brandtbucher** [comment]
Congrats on your first CPython contribution @curtisbucher! :champagne:

Looking forward to seeing more from you in the future.

**Comment by gvanrossum** [comment]
Well, we missed 3.9.0a5. Oh well.

**Comment by gvanrossum** [comment]
Calling out @rhettinger -- if you approve of the version in https://github.com/python/cpython/pull/18832#issuecomment-596910350 please say so and well make it happen.

**Pull request merged**
